### ISSUE TITLE
Stack overflow after migration to 10.0.1 from 9.0.0

### ISSUE THREAD

**Issue by plokhotnyuk** [julienrf/play-json-derived-codecs#84]
Here is a code snippet that works fine using 9.0.0 but fails with stack overflow using 10.0.1:
```
import julienrf.json.derived.flat
import play.api.libs.functional.syntax._
import play.api.libs.json._

sealed trait ADTBase extends Product with Serializable
case class X(a: Int) extends ADTBase
case class Y(b: String) extends ADTBase
case class Z(l: ADTBase, r: ADTBase) extends ADTBase

val adtFormat: Format[ADTBase] = {
  implicit lazy val v1: Format[X] = Json.format
  implicit lazy val v2: Format[Y] = Json.format
  implicit lazy val v3: Format[Z] = Json.format
  implicit lazy val v4: Format[ADTBase] = flat.oformat((__ \ "type").format[String])
  v4
}

println(Json.parse("""{"type":"Z","l":{"type":"X","a":1},"r":{"type":"Y","b":"VVV"}}""").as[ADTBase](adtFormat))
```
Link to Scastie: https://scastie.scala-lang.org/WIYG5my6TNyIYF8zsWIkCg

**Comment by julienrf** [comment]
@ncreep would you be interested in having a look?

**Comment by ncreep** [comment]
Recursion my old friend...

Will take a look.

**Comment by ncreep** [comment]
Opened a pull request with the fix: #85.
